### DISCUSSION TITLE
Why would we need to differentiate through data augmentations? (i.e., why kornia)

### DISCUSSION THREAD

**Discussion by vedantroy** [kornia/kornia#1829] [Questions]
Hi!

This is a potentially very silly question, but why would we need the ability to differentiate *through* data augmentations?
Presumably back-propagation ends before the augmentation layer.
E.g, is there a reason we would want to / need to differentiate through, e.g, a `RandomCrop`?

An example is here: https://github.com/lucidrains/imagen-pytorch/blob/9c14f125bb42d90974552fb15cf6e24b213d8a4e/imagen_pytorch/imagen_pytorch.py#L2286.

Cross-posted on pytorch forums: https://discuss.pytorch.org/t/why-would-we-need-to-differentiate-through-data-augmentations/159378

**Comment by catalys1** [comment]
1. Learnable augmentations; that is, you could parameterize the augmentations and then learn the parameters along with everything else.
2. Sometimes you actually want to backpropogate into image space; here's [an example](https://distill.pub/2018/differentiable-parameterizations/)

That's just two off the top of my head. I'm sure there are plenty of others. Take a look at the [Kornia paper](https://arxiv.org/abs/1910.02190).

Of course, Kornia isn't just for differentiability either. It has a lot of vision operations implemented directly in pytorch, which is useful even if you don't need to differentiate through them.
